### ISSUE TITLE
403 response from S3 despite valid credentials being present in environmental variables

### ISSUE THREAD

**Issue by mfitz** [RaRe-Technologies/smart_open#136]
Hi

I'm trying to stream the content of an object in my S3 bucket using a very trivial app:

```
import smart_open
import sys


if __name__ == '__main__':
    s3_url_to_read = sys.argv[1]

    for line in smart_open.smart_open(s3_url_to_read):
        print(line)
```

I have valid AWS credentials in the relevant environmental variables within the shell, which the aws CLI and other python apps based on boto3 are able to detect and use without issue; I can list objects in the bucket via `aws s3 ls`, for example. However, when I run the trivial app above in the same shell, I get a 403 error:

```
Traceback (most recent call last):
  File "/Users/mickyfitz/workspace/api-harvester/tools/datalake/smart_open_test.py", line 9, in <module>
    for line in smart_open.smart_open(s3_url_to_read):
  File "/private/tmp/big-file-grabber-experiments/2-7-venv/lib/python2.7/site-packages/smart_open/smart_open_lib.py", line 163, in smart_open
    bucket = s3_connection.get_bucket(parsed_uri.bucket_id)
  File "/private/tmp/big-file-grabber-experiments/2-7-venv/lib/python2.7/site-packages/boto/s3/connection.py", line 509, in get_bucket
    return self.head_bucket(bucket_name, headers=headers)
  File "/private/tmp/big-file-grabber-experiments/2-7-venv/lib/python2.7/site-packages/boto/s3/connection.py", line 542, in head_bucket
    raise err
boto.exception.S3ResponseError: S3ResponseError: 403 Forbidden
```

I see the exact same error using both python 2.7 and python 3.6. Am I missing something? Have you seen this behaviour before? I believe smart_open is based on the boto library rather than boto3, but I expected my key and secret to be picked up from env vars in the same way they are with boto3. Could this be related in some way to missing policies on the bucket?

**Comment by mfitz** [comment]
I see the same behaviour with buckets that have public permissions, for example `s3://us-east-1.elasticmapreduce.samples/flightdata/sparkapp/FlightExample.scala`, although I did have to work around an SSL certs problem first, meaning the app now looks like:

```
import smart_open
import sys

import ssl

_old_match_hostname = ssl.match_hostname


def _new_match_hostname(cert, hostname):
    if hostname.endswith('.s3.amazonaws.com'):
        pos = hostname.find('.s3.amazonaws.com')
        hostname = hostname[:pos].replace('.', '') + hostname[pos:]
    return _old_match_hostname(cert, hostname)

ssl.match_hostname = _new_match_hostname


if __name__ == '__main__':
    s3_url_to_read = sys.argv[1]

    for line in smart_open.smart_open(s3_url_to_read):
        print(line)
```

When I run this, I see the same error:

```
python smart_open_test.py s3://us-east-1.elasticmapreduce.samples/flightdata/sparkapp/FlightExample.scala

Traceback (most recent call last):
  File "/Users/mickyfitz/workspace/api-harvester/tools/datalake/smart_open_test.py", line 20, in <module>
    for line in smart_open.smart_open(s3_url_to_read):
  File "/private/tmp/big-file-grabber-experiments/2-7-venv/lib/python2.7/site-packages/smart_open/smart_open_lib.py", line 163, in smart_open
    bucket = s3_connection.get_bucket(parsed_uri.bucket_id)
  File "/private/tmp/big-file-grabber-experiments/2-7-venv/lib/python2.7/site-packages/boto/s3/connection.py", line 509, in get_bucket
    return self.head_bucket(bucket_name, headers=headers)
  File "/private/tmp/big-file-grabber-experiments/2-7-venv/lib/python2.7/site-packages/boto/s3/connection.py", line 542, in head_bucket
    raise err
boto.exception.S3ResponseError: S3ResponseError: 403 Forbidden
```

**Comment by mpenkov** [comment]
@mfitz Thanks for reporting this. Could you please try to reproduce the problem with the current master? It uses boto3 now, which may influence how the credentials are picked up.

**Comment by mfitz** [comment]
Hi @mpenkov 

Thanks for your reply - I can confirm that this is indeed fixed using the current master. Excellent news!

**Comment by menshikh-iv** [comment]
Great @mfitz, thanks for the confirmation!

I close this, because this bug already fixed in current master fbc82cc04bf92e8c588c49c43ff5aff8234ea87d
